include/ncutility/SpscQueue.h

```diff
@@ -0,0 +1,109 @@
+/**
+ * @file SpscQueue.h
+ * @copyright Jaremie Romer and McCallister Romer 2025
+ */
+#pragma once
+
+#include "ncutility/platform/New.h"
+
+#include <array>
+#include <atomic>
+
+namespace nc
+{
+NC_DISABLE_WARNING_PUSH
+NC_DISABLE_WARNING_MSVC(4324)
+
+/** @brief Fixed-size single-producer/single-consumer queue. */
+template <class T, size_t N>
+class spsc_queue
+{
+    public:
+        [[nodiscard]] auto push(const T& v) -> bool
+        {
+            return emplace(v);
+        }
+
+        [[nodiscard]] auto push(T&& v) -> bool
+        {
+            return emplace(std::move(v));
+        }
+
+        [[nodiscard]] auto pop(T& out) -> bool
+        {
+            if (is_empty(m_consumer))
+            {
+                refresh_consumer_cache();
+                if (is_empty(m_consumer))
+                {
+                    return false;
+                }
+            }
+
+            out = std::move(m_buffer[index(m_consumer.head)]);
+            advance_consumer();
+            return true;
+        }
+
+        template<class... Args>
+        [[nodiscard]] auto emplace(Args&&... args) -> bool
+        {
+            if (is_full(m_producer))
+            {
+                refresh_producer_cache();
+                if (is_full(m_producer))
+                {
+                    return false;
+                }
+            }
+
+            m_buffer[index(m_producer.tail)] = T(std::forward<Args>(args)...);
```

Comment:
What happens to the T that was in `m_buffer[index(m_producer.tail)]` here? Is this guaranteed to only run if `m_buffer[index(m_producer.tail)]` is empty so it doesn't overwrite the previous T?

Reply:
The T that was previously here must have been already popped before we got here. AKA `pop()` is guaranteed to move from this buffer location before we can write back to it.